filebeat/input/filestream/internal/input-logfile/harvester.go

```diff
@@ -110,12 +132,27 @@ type defaultHarvesterGroup struct {
 
 // Start starts the Harvester for a Source. It does not block.
 func (hg *defaultHarvesterGroup) Start(ctx input.Context, s Source) {
-	sourceName := s.Name()
-
-	ctx.Logger = ctx.Logger.With("source", sourceName)
+	ctx.Logger = ctx.Logger.With("source", s.Name())
 	ctx.Logger.Debug("Starting harvester for file")
 
-	hg.tg.Go(func(canceler unison.Canceler) error {
+	hg.tg.Go(startHarvester(ctx, hg, s, 0))
+}
+
+// Restart starts the Harvester for a Source if a Harvester is already running it waits for it
+// to shut down for a specified timeout. It does not block.
+func (hg *defaultHarvesterGroup) Restart(ctx input.Context, s Source) {
+
+	ctx.Logger = ctx.Logger.With("source", s.Name())
+	ctx.Logger.Debug("Restarting harvester for file")
+
+	// we are waiting for five seconds so harvester has enough time to shut down
+	hg.tg.Go(startHarvester(ctx, hg, s, 5*time.Second))
```

Comment:
Do we really want to fail? What will happen in that case? Will the scan based prospector eventually restart the harvester again? What if we make an inotify based prospector, then we get no second chance?

The stateStore was designed to block on purpose, such that we can immediately continue when the original harvester did return, independent of the duration it takes to release the state (early cancellation is still possible).

Reply:
No, I had revisited my inital reasoning.